File: MinStack.java
import java.util.ArrayList;
import java.util.List;

public class MinStack {
    private List<Integer> stack;
    private List<Integer> minStack;
    
    public MinStack() {
        stack = new ArrayList<>();
        minStack = new ArrayList<>();
    }
    
    public void push(int val) {
        stack.add(val);
        if(minStack.size()==0){
            minStack.add(val);
            return;
        }

        int min = minStack.get(minStack.size()-1);
        minStack.add(val>min ? min : val);
    }
    
    
    public void pop() {
        int length = stack.size()-1;
        if(length >= 0){
            stack.remove(length);
            minStack.remove(length);
        }    
    }
    
    public int top() {
        if(stack.size() != 0)
            return stack.getLast();
        return 0;    
    }
    
    public int getMin() {
        if(minStack.size() > 0)
            return minStack.get(minStack.size()-1);
        return 0;    
    }

    public static void main(String[] args) {
        MinStack minStack = new MinStack();
        minStack.push(-2);
        minStack.push(0);
        minStack.push(-3);
        System.out.println(minStack.getMin());
        minStack.pop();
        System.out.println(minStack.top());
        System.out.println(minStack.getMin());
    }
}

File: minStack.go
package main

type MinStack struct {
	array []int
	min   []int
}

func Constructor() MinStack {
	return MinStack{make([]int, 0), make([]int, 0)}
}

func (this *MinStack) Push(val int) {
	if len(this.min) == 0 {
		this.min = append(this.min, val)
		this.array = append(this.array, val)
	} else {
		this.min = append(this.min, min(this.min[len(this.min)-1], val))
		this.array = append(this.array, val)
	}

}

func (this *MinStack) Pop() {
	l := len(this.min) - 1
	this.min = this.min[:l]
	this.array = this.array[:l]
}

func (this *MinStack) Top() int {
	return this.array[len(this.array)-1]
}

func (this *MinStack) GetMin() int {
	return this.min[len(this.min)-1]
}
